### PR TITLE
Add the diffusers examples of inference Tech

### PR DESCRIPTION
**Pull request by yuanwu2017**
# What does this PR do?
This PR includes three examples of  inference techniques.
1. Control brightness
2. Prompt weighting
3. Improve the image quality with Freeu

Known issues:
Currently HPU cannot support the torch.fft operations, so running these operations on CPU.

@regisss @libinta @mandy-li 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: examples/stable-diffusion/README.md [hunk 1]
```diff
@@ -584,3 +584,66 @@ python image_to_video_generation.py \
...
+Here is how to run the diffusers examples of inference techniques. For more details,
+please refer to [Hugging Face Diffusers doc](https://huggingface.co/docs/diffusers/main/en/using-diffusers/overview_techniques).
+
+## Control brightness
+
+Here is how to run the example of controlling brightness. For more details,
+please refer to [Hugging Face Diffusers doc](https://huggingface.co/docs/diffusers/main/en/using-diffusers/control_brightness).
+
+```bash
+python text_to_image_generation.py \
```
Comment (vidyasiv) [inline examples/stable-diffusion/README.md:599]: I see an error no module named compel, please update examples/stable-diffusion/requirements.txt

**Review by vidyasiv — CHANGES_REQUESTED** [review]
Thank you for your PR @yuanwu2017, please update requirements.txt as indicated

File: optimum/habana/diffusers/models/unet_2d_condition.py [hunk 2]
```diff
@@ -4,12 +4,44 @@
...
+def gaudi_fourier_filter(x_in: "torch.Tensor", threshold: int, scale: int) -> "torch.Tensor":
+    """Fourier filter as introduced in FreeU (https://arxiv.org/abs/2309.11497).
+
+    This version of the method comes from here:
+    https://github.com/huggingface/diffusers/pull/5164#issuecomment-1732638706
+    """
+    x = x_in
+    B, C, H, W = x.shape
+
+    # FFT
```
Comment (vidyasiv) [inline optimum/habana/diffusers/models/unet_2d_condition.py:28]: Could you add a comment here: "Moving to CPU as torch.fft operations are not supported on HPU"

File: examples/stable-diffusion/README.md [hunk 3]
```diff
@@ -584,3 +584,66 @@ python image_to_video_generation.py \
...
+    --timestep_spacing trailing
+```
+
+## Prompt weighting
+
+Here is how to run the example of prompt weighting. For more details,
+please refer to [Hugging Face Diffusers doc](https://huggingface.co/docs/diffusers/main/en/using-diffusers/weighted_prompts).
+
+```bash
+pip install compel
```
Comment (vidyasiv) [inline examples/stable-diffusion/README.md:619]: this line can be removed if we add to requirements.txt

File: examples/stable-diffusion/README.md (hunk 1 above)
Comment (yuanwu2017) [reply examples/stable-diffusion/README.md:599]: Done.

File: examples/stable-diffusion/README.md [hunk 4]
```diff
@@ -584,3 +584,66 @@ python image_to_video_generation.py \
...
+    --timestep_spacing trailing
+```
+
+## Prompt weighting
+
+Here is how to run the example of prompt weighting. For more details,
+please refer to [Hugging Face Diffusers doc](https://huggingface.co/docs/diffusers/main/en/using-diffusers/weighted_prompts).
+
+```bash
+pip install -r requirements.txt
```
Comment (vidyasiv) [inline examples/stable-diffusion/README.md:619]:
```suggestion

```

**Review by vidyasiv — CHANGES_REQUESTED** [review]

File: examples/stable-diffusion/README.md [hunk 5]
```diff
@@ -584,3 +584,66 @@ python image_to_video_generation.py \
...
+```
+
+## Controlling image quality
+
+Here is how to run the example of improving image quality. For more details,
+please refer to [Hugging Face Diffusers doc](https://huggingface.co/docs/diffusers/main/en/using-diffusers/image_quality).
+
+```bash
+python text_to_image_generation.py \
+    --model_name_or_path runwayml/stable-diffusion-v1-5 \
```
Comment (vidyasiv) [inline examples/stable-diffusion/README.md:640]:
```suggestion
    --model_name_or_path CompVis/stable-diffusion-v1-4 \
```

**Review by vidyasiv — CHANGES_REQUESTED** [review]
Couple minor changes

**Review by vidyasiv — APPROVED** [review]
small change needed

**Review by vidyasiv — CHANGES_REQUESTED** [review]
small change needed

**Comment by yuanwu2017** [comment]
> small change needed

What small changes are needed?

**Review by vidyasiv — APPROVED** [review]

**Comment by vidyasiv** [comment]
@regisss  please take a look

File: optimum/habana/diffusers/models/unet_2d_condition.py [hunk 6]
```diff
@@ -4,12 +4,45 @@
...
+    B, C, H, W = x_freq.shape
+    mask = torch.ones((B, C, H, W), device=x.device)
+    crow, ccol = H // 2, W // 2
+    mask[..., crow - threshold : crow + threshold, ccol - threshold : ccol + threshold] = scale
+    x_freq = x_freq * mask
+
+    # IFFT
+    x_freq = ifftshift(x_freq, dim=(-2, -1))
+    x_filtered = ifftn(x_freq, dim=(-2, -1)).real
+    return x_filtered.to(device=x_in.device, dtype=x_in.dtype)
```
Comment (regisss) [inline optimum/habana/diffusers/models/unet_2d_condition.py:43]: Can you move this to a new file in `optimum/habana/diffusers/utils/torch_utils.py` and add a link to the [reference](https://github.com/huggingface/diffusers/blob/v0.29.2/src/diffusers/utils/torch_utils.py#L93) in the docstring please? It will be easier to maintain that way.

File: examples/stable-diffusion/text_to_image_generation.py [hunk 7]
```diff
@@ -270,6 +271,21 @@ def main():
...
+    )
+    parser.add_argument(
+        "--use_freeu",
+        action="store_true",
+        help="Use freeu for improving generation quality",
+    )
+    parser.add_argument(
+        "--use_zero_snr",
+        action="store_true",
+        help="Use rescale_betas_zero_snr for control image brightness",
```
Comment (regisss) [inline examples/stable-diffusion/text_to_image_generation.py:287]:
```suggestion
        help="Use rescale_betas_zero_snr for controlling image brightness",
```

File: examples/stable-diffusion/README.md [hunk 8]
```diff
@@ -584,3 +584,65 @@ python image_to_video_generation.py \
 
 > For improved performance of the image-to-video pipeline on Gaudi, it is recommended to configure the environment
 > by setting PT_HPU_MAX_COMPOUND_OP_SIZE to 1.
+
+# Diffusers Inference techniques
```
Comment (regisss) [inline examples/stable-diffusion/README.md:588]: I wouldn't set it here with this section name.
Rather `## Additional inference techniques` that could be added right before https://github.com/yuanwu2017/optimum-habana/tree/tech/examples/stable-diffusion#stable-video-diffusion-examples.

File: examples/stable-diffusion/README.md [hunk 9]
```diff
@@ -584,3 +584,65 @@ python image_to_video_generation.py \
...
+
+Here is how to run the diffusers examples of inference techniques. For more details,
+please refer to [Hugging Face Diffusers doc](https://huggingface.co/docs/diffusers/main/en/using-diffusers/overview_techniques).
+
+## Control brightness
```
Comment (regisss) [inline examples/stable-diffusion/README.md:593]:
```suggestion
## Controlling brightness
```

**Comment by vidyasiv** [comment]
@yuanwu2017 , please run make style

**Review by vidyasiv — APPROVED** [review]

File: notebooks/AI_HW_Summit_2022.ipynb
Comment (regisss) [inline notebooks/AI_HW_Summit_2022.ipynb:1]: Why these changes? Were they brought by `make style`?
